### ISSUE TITLE
v2.3.5 fails on delete with server error 500

### ISSUE THREAD

**Issue by ceecko** [coreos/etcd#6115]
We're running a 3 member cluster. We usually set values in batches - i.e. we set ~15 values serially, one after another, through http API.

13 out of 15 values are set using HTTP PUT and 2/15 are removed using HTTP DELETE.
During one batch the 13 PUT values are set without errors but sometimes the one of the DELETE operations fails with server error 500.

On one of the members the log says:

```
etcd[1100]: got unexpected response error (etcdserver: request timed out)
```

This isn't easily reproducible, but happens from time to time.
Is there anything special about DELETE which should cause any issues?

If it matters, we're using https to access the cluster.

**Comment by heyitsanthony** [comment]
@ceecko do you have etcd server logs from around the time the requests time out? It's possible that the cluster is unstable / underspec causing it to occasionally lose the leader and timeout on requests.

**Comment by ceecko** [comment]
@heyitsanthony there's actually nothing around the time. Here's to log before and after
```
Aug 05 11:24:33 etcd.localdomain etcd[1100]: purged file /data/etcd/member/snap/0000000000000269-0000000000f62698.snap successfully
Aug 05 11:46:59 etcd.localdomain etcd[1100]: got unexpected response error (etcdserver: request timed out)
Aug 05 11:47:09 etcd.localdomain etcd[1100]: got unexpected response error (etcdserver: request timed out)
Aug 05 12:39:15 etcd.localdomain etcd[1100]: start to snapshot (applied: 16191742, lastsnap: 16181741)
```

**Comment by heyitsanthony** [comment]
@ceecko OK, the cluster itself seems healthy. I suspect the problem is related to the v2 store.

Some follow up questions:
* Is the delete removing a directory?
* Do the timed out requests delete the data on the cluster or is the data still there?
* What is the typical latency for a request (e.g., `etcdctl get -q somekey`)?
* Does the cluster service many concurrent requests at once?
* How many keys are on the cluster / how big is the data set?
* Is this new behavior?

**Comment by ceecko** [comment]
- Is the delete removing a directory?  
No it's a key
- Do the timed out requests delete the data on the cluster or is the data still there?  
Haven't checked, but when the issue occurs, I'll check.
- What is the typical latency for a request (e.g., etcdctl get -q somekey)?  
It's pretty quick, 50-150ms
- Does the cluster service many concurrent requests at once?  
There's not high load on etcd in general. When the issue occurs, usually there's only these 13 PUT requests and 2 DELETE, all serially with 4 watchers.
- How many keys are on the cluster / how big is the data set?  
I don't know how to find the size of the data set, but it's roughly 4k keys each a couple bytes (short string)
- Is this new behavior?  
This has happened with older versions as well, but didn't really have the time to investigate.

**Comment by xiang90** [comment]
@ceecko Can you please provide etcd server log across the cluster for a 10 minutes window when the 500 error happens? I feel it is caused by the unstable cluster.

**Comment by xiang90** [comment]
Also are you able to reliably reproduce this issue with a simple setup on AWS or GCE or other env that we can get? Hopefully we can reproduce it and help you dig into the problem more effectively.

**Comment by ceecko** [comment]
@xiang90 here are the log files from each member.
I can't easily reproduce the error though. We had to change the code so that keys are not deleted, so we're not experiencing the error anymore, but I'll try to create a test env to reproduce this and let you know.


```
Aug 05 10:06:05 member1 etcd[13911]: start to snapshot (applied: 16172349, lastsnap: 16162348)
Aug 05 10:06:05 member1 etcd[13911]: saved snapshot at index 16172349
Aug 05 10:06:05 member1 etcd[13911]: compacted raft log at 16167349
Aug 05 10:06:26 member1 etcd[13911]: purged file /data/etcd/member/snap/0000000000000269-0000000000f601e8.snap successfully
Aug 05 11:29:09 member1 etcd[13911]: start to snapshot (applied: 16182350, lastsnap: 16172349)
Aug 05 11:29:09 member1 etcd[13911]: saved snapshot at index 16182350
Aug 05 11:29:09 member1 etcd[13911]: compacted raft log at 16177350
Aug 05 11:29:26 member1 etcd[13911]: purged file /data/etcd/member/snap/0000000000000269-0000000000f628f9.snap successfully
Aug 05 12:44:20 member1 etcd[13911]: start to snapshot (applied: 16192351, lastsnap: 16182350)
Aug 05 12:44:20 member1 etcd[13911]: saved snapshot at index 16192351
Aug 05 12:44:20 member1 etcd[13911]: compacted raft log at 16187351
Aug 05 12:44:26 member1 etcd[13911]: purged file /data/etcd/member/snap/0000000000000269-0000000000f6500a.snap successfully
Aug 05 14:07:01 member1 etcd[13911]: start to snapshot (applied: 16202352, lastsnap: 16192351)
Aug 05 14:07:01 member1 etcd[13911]: saved snapshot at index 16202352
Aug 05 14:07:01 member1 etcd[13911]: compacted raft log at 16197352
Aug 05 14:07:26 member1 etcd[13911]: purged file /data/etcd/member/snap/0000000000000269-0000000000f6771b.snap successfully
```


```
Aug 05 08:39:16 member2 etcd[1100]: start to snapshot (applied: 16161739, lastsnap: 16151738)
Aug 05 08:39:16 member2 etcd[1100]: saved snapshot at index 16161739
Aug 05 08:39:16 member2 etcd[1100]: compacted raft log at 16156739
Aug 05 08:39:32 member2 etcd[1100]: purged file /data/etcd/member/snap/0000000000000269-0000000000f5d876.snap successfully
Aug 05 10:01:04 member2 etcd[1100]: start to snapshot (applied: 16171740, lastsnap: 16161739)
Aug 05 10:01:04 member2 etcd[1100]: saved snapshot at index 16171740
Aug 05 10:01:04 member2 etcd[1100]: compacted raft log at 16166740
Aug 05 10:01:32 member2 etcd[1100]: purged file /data/etcd/member/snap/0000000000000269-0000000000f5ff87.snap successfully
Aug 05 11:18:57 member2 etcd[1100]: got unexpected response error (etcdserver: request timed out)
Aug 05 11:24:04 member2 etcd[1100]: start to snapshot (applied: 16181741, lastsnap: 16171740)
Aug 05 11:24:04 member2 etcd[1100]: saved snapshot at index 16181741
Aug 05 11:24:04 member2 etcd[1100]: compacted raft log at 16176741
Aug 05 11:24:33 member2 etcd[1100]: purged file /data/etcd/member/snap/0000000000000269-0000000000f62698.snap successfully
Aug 05 11:46:59 member2 etcd[1100]: got unexpected response error (etcdserver: request timed out)
Aug 05 11:47:09 member2 etcd[1100]: got unexpected response error (etcdserver: request timed out)
Aug 05 12:39:15 member2 etcd[1100]: start to snapshot (applied: 16191742, lastsnap: 16181741)
Aug 05 12:39:15 member2 etcd[1100]: saved snapshot at index 16191742
Aug 05 12:39:15 member2 etcd[1100]: compacted raft log at 16186742
Aug 05 12:39:33 member2 etcd[1100]: purged file /data/etcd/member/snap/0000000000000269-0000000000f64da9.snap successfully
Aug 05 14:01:57 member2 etcd[1100]: start to snapshot (applied: 16201743, lastsnap: 16191742)
Aug 05 14:01:57 member2 etcd[1100]: saved snapshot at index 16201743
Aug 05 14:01:57 member2 etcd[1100]: compacted raft log at 16196743
Aug 05 14:02:03 member2 etcd[1100]: purged file /data/etcd/member/snap/0000000000000269-0000000000f674ba.snap successfully
```
```
Aug 05 10:00:12 member3 etcd[28500]: start to snapshot (applied: 16171635, lastsnap: 16161634)
Aug 05 10:00:12 member3 etcd[28500]: saved snapshot at index 16171635
Aug 05 10:00:12 member3 etcd[28500]: compacted raft log at 16166635
Aug 05 10:00:39 member3 etcd[28500]: purged file /data/etcd/member/snap/0000000000000269-0000000000f5ff1e.snap successfully
Aug 05 11:18:10 member3 etcd[28500]: failed to send out heartbeat on time (deadline exceeded for 148.742785ms)
Aug 05 11:18:10 member3 etcd[28500]: server is likely overloaded
Aug 05 11:18:10 member3 etcd[28500]: failed to send out heartbeat on time (deadline exceeded for 159.253454ms)
Aug 05 11:18:10 member3 etcd[28500]: server is likely overloaded
Aug 05 11:18:52 member3 etcd[28500]: failed to send out heartbeat on time (deadline exceeded for 20.611643ms)
Aug 05 11:18:52 member3 etcd[28500]: server is likely overloaded
Aug 05 11:18:52 member3 etcd[28500]: failed to send out heartbeat on time (deadline exceeded for 20.674513ms)
Aug 05 11:18:52 member3 etcd[28500]: server is likely overloaded
Aug 05 11:23:12 member3 etcd[28500]: start to snapshot (applied: 16181636, lastsnap: 16171635)
Aug 05 11:23:12 member3 etcd[28500]: saved snapshot at index 16181636
Aug 05 11:23:12 member3 etcd[28500]: compacted raft log at 16176636
Aug 05 11:23:40 member3 etcd[28500]: purged file /data/etcd/member/snap/0000000000000269-0000000000f6262f.snap successfully
Aug 05 12:38:23 member3 etcd[28500]: start to snapshot (applied: 16191637, lastsnap: 16181636)
Aug 05 12:38:23 member3 etcd[28500]: saved snapshot at index 16191637
Aug 05 12:38:23 member3 etcd[28500]: compacted raft log at 16186637
Aug 05 12:38:40 member3 etcd[28500]: purged file /data/etcd/member/snap/0000000000000269-0000000000f64d40.snap successfully
Aug 05 14:01:05 member3 etcd[28500]: start to snapshot (applied: 16201638, lastsnap: 16191637)
Aug 05 14:01:05 member3 etcd[28500]: saved snapshot at index 16201638
Aug 05 14:01:05 member3 etcd[28500]: compacted raft log at 16196638
Aug 05 14:01:10 member3 etcd[28500]: purged file /data/etcd/member/snap/0000000000000269-0000000000f67451.snap successfully
```

**Comment by xiang90** [comment]
@ceecko I could not see anything from the log. Can you try to get metrics from all members by accessing HTTP endpoint at clienturl/metrics? It would be great if you can get a reproduce.

**Comment by ceecko** [comment]
Attached are the metrics from all three members.
I have created a script which replicates the same behavior, but since it's so unpredictable when the error occurs, it may take weeks until I can replicate it and it's not going to be reliable. In case the error occurs I'm going to check if the key has been deleted or not. 

[m1.txt](https://github.com/coreos/etcd/files/425775/m1.txt)
[m2.txt](https://github.com/coreos/etcd/files/425779/m2.txt)
[m3.txt](https://github.com/coreos/etcd/files/425780/m3.txt)

**Comment by xiang90** [comment]
@ceecko Thanks for providing the metrics.

I gave them a close look, and figured out the root cause of the timeout.

If you look at the metrics, you can find quite a few fsyncs that took more than 4 seconds:

m1 -> 1 fsyncs took more than 4 seconds (le8.192) - le(4.096) = 1
```
etcd_wal_fsync_durations_seconds_bucket{le="4.096"} 1.6644482e+07
etcd_wal_fsync_durations_seconds_bucket{le="8.192"} 1.6644483e+07
```

m2 -> 1000s fsyncs took more than 4 seconds
```
etcd_wal_fsync_durations_seconds_bucket{le="4.096"} 3.369961e+06
etcd_wal_fsync_durations_seconds_bucket{le="8.192"} 3.370211e+06
etcd_wal_fsync_durations_seconds_bucket{le="+Inf"} 3.370286e+06
```

m3 -> 100s fsyncs took more than 4 seconds

```
etcd_wal_fsync_durations_seconds_bucket{le="4.096"} 2.98e+06
etcd_wal_fsync_durations_seconds_bucket{le="8.192"} 2.980263e+06
etcd_wal_fsync_durations_seconds_bucket{le="+Inf"} 2.980287e+06
```

Once fsync takes more than 5 seconds, there will be a timeout.

Also not only delete, but also put timed out.

for example:
```
etcd_http_failed_total{code="500",method="DELETE"} 5
etcd_http_failed_total{code="500",method="POST"} 1
etcd_http_failed_total{code="500",method="PUT"} 9
```

In summary, the machines you run etcd cluster have unreliable disk IO. etcd is very sensitive to disk performance. If you upgrade to etcd 3+, etcd will print out warnings if fsync takes too long.

As we figured out the root cause, I close this issue.
